File: languages/ocaml/fib.ml
let rec fibrec n = match n with
  | 1 -> 1
  | 2 -> 1
  | _ -> fibrec (n-1) + fibrec (n-2)

let fib n =
  let rec recur n a b = (match n with
    | 1 -> a
    | _ -> recur (n-1) b (a+b))
  in recur n 1 1

let rec do_between f low high = match low <= high with
  | true -> f low; do_between f (low+1) high
  | false -> ()

let main =
  do_between (fun n -> p_int (fibrec n)) 1 10;
  do_between (fun n -> p_int (fib n)) 1 10

File: languages/lua/fib.lua
function fibrec(n)
  if n == 0 then
    return 1
  elseif n == 1 then
    return 1
  end
  return fibrec(n-1) + fibrec(n-2)
end

function fib(n)
  local a = 0
  local b = 1
  local temp
  for i=1,n-1 do
    temp = b
    b = a + b
    a = temp
  end
  return b
end

print("rec")
for i=1,10 do
  print(fibrec(i))
end

print("non-rec")
for i=20,50 do
  print(fib(i))
end
